sunpy/timeseries/metadata.py

```diff
@@ -0,0 +1,433 @@
+__email__ = "stuart@mumford.me.uk"
+
+from sunpy.util.metadata import MetaDict
+
+from sunpy.time import TimeRange, parse_time
+
+class TimeSeriesMetaData:
+    """
+    An object used to store metadata for TimeSeries objects that enables multiple
+    TimeSeries metadata to be concatenated in an organised fashion.
+
+    Attributes
+    ----------
+    metadata : `list` of `tuple`
+        The list of 3-tuples which each represent a source files metadata.
+        The tuples consist of: ( TimeRange, [ colnames ], MetaDict(metadata) )
+
+    Examples
+    --------
+    >>> from sunpy.timeseries import TimeSeriesMetaData
+    >>> from sunpy.time import TimeRange, parse_time
+    >>> from collections import MetaDict
+    >>> tr = TimeRange('2012-06-01 00:00','2012-06-02 00:00')
+    >>> md = TimeSeriesMetaData()
+    >>> md.append(tr, ['GOES'], MetaDict([('tr','tr')]))
+    >>> md.find(parse_time('2012-06-01T21:08:12'))
+    >>> md.find(parse_time('2012-06-01T21:08:12'), 'GOES')   # doctest: +SKIP
+    """
+
+    def __init__(self, meta=None, timerange=None, colnames=None, **kwargs):
+        self.metadata = []
+        if meta:
+            if isinstance(meta, (dict, MetaDict)) and isinstance(timerange, TimeRange) and isinstance(colnames, list):
+                # Given a single metadata entry as a dictionary with additional timerange and colnames.
+                self.metadata.append((timerange, colnames, meta))
+            elif isinstance(meta, tuple):
+                # Given a single metadata entry as a tuple.
+                self.metadata.append(meta)
+            elif isinstance(meta, list):
+                # Given a complex metadata list (of tuples)
+                self.metadata = meta.copy()
+
+    def append(self, timerange, columns, metadata, **kwargs):
+        """
+        Add the given metadata MetaDict into the metadata list as a tuple with
+        it's TimeRange and colnames (list).
+        Will add the new entry so the list is in chronological order for the
+        TimeRange.start datetime values.
+        
+        Parameters
+        ----------
+        timerange : `~sunpy.time.TimeRange`
+            The timerange for which a given metadict is relevant. This will
+            generally initilly be the full range of the original file, but if
+            the TimeSeries gets truncated this may change appropriately.
+        
+        columns : `str`
+            A list of the colomn name strings that the metadata is relevant for.
+
+        metadata : `~sunpy.util.metadata.MetaDict` or `OrderedDict` or `dict`
+            The dictionary holding the metadata.
+        """
+        # Parameters
+        metadata = MetaDict(metadata)
+        
+        # Check the types are correct.
+        pos = len(self.metadata)
+        if isinstance(timerange, TimeRange):
+            for i, meta in enumerate(self.metadata):
+                if timerange.start < meta[0].start:
+                    pos = i - 1

```

Comment:
I haven't actually run this, but this logic doesn't read right to me:
* This really shouldn't be `i - 1` because you'll get the wrong behavior if `i == 0` satisfies the logical criterion, and you could never get `pos == len(self.metadata) - 1`, which should be a valid insert position.  Note that `insert()` inserts before the existing element at the specified position.
* The logical criterion, combined with the lack of a `break`, seems to be structured wrong.  You end up settling on the last element that is after `timerange.start`, rather than the first element.  This would be more cleanly written as a `while` loop, rather than the `enumerate` approach.

Try testing this with longer lists for `self.metadata`, because degeneracies when testing on one- or two-element lists may obscure such bugs.